//
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

edition = "2023";

// These protos are currently used to store structured output for pipe DESCRIBE.
// That output is not exposed publicly yet so these protos are still private.
// When structured output from DESCRIBE or other operators becomes public,
// these protos should move to ../public/proto.
package zetasql.reflection;

option java_multiple_files = true;

// This describes a column as part of a table schema.
message Column {
  // The table alias for this column, if this column has one unique table alias.
  string table_alias = 1;

  string column_name = 2;

  // The column's type, formatted as a string.
  string type = 3;

  // Indicates this column is a value table column. Its top-level fields will
  // be implicitly selectable without using the column name.
  // It's possible to have multiple columns in a result, where some are value
  // table columns.  e.g. This occurs after using JOIN or EXTEND on value
  // tables.
  bool is_value_table_column = 4;
}

// This describes a table alias (also known as a range variable),
// with the list of column names available under that alias.
// This is also used to describe CTEs.
message TableAlias {
  string name = 1;
  repeated string column_name = 2;
  repeated string pseudo_column_name = 3;
}

// This describes an intermediate result table.
// This is produced by pipe DESCRIBE.
message ResultTable {
  // This list of visible top-level columns. This is what will be shown in the
  // result table if this table is returned as the final query output.
  repeated Column column = 1;

  // The list of other column-like names that are in scope, but aren't part of
  // the visible column-list.  This includes pseudo-columns, which are
  // selectable by name but aren't included in `SELECT *`.
  repeated Column pseudo_column = 2;

  // The list of table aliases (range variables) currently in scope.
  repeated TableAlias table_alias = 3;

  // The list of CTEs (names from WITH clauses) visible at this point in the
  // query.
  repeated TableAlias common_table_expression = 4;

  // Indicates this intermediate table is a value table.
  // If true, the `column` list will always have exactly one column, where that
  // column's type is the row type, and its value is the row value.
  bool is_value_table = 5;

  // Indicates this intermediate result table is ordered.
  // This matches `is_ordered` on the `ResolvedScan` preceding this DESCRIBE.
  bool is_ordered = 6;
}
